### PR TITLE
[# 52] Feature / New design DistributedAtomSpace

### PR DESCRIPTION
**Pull request by marcocapozzoli**

### CODE REVIEW

File: hyperon_das/client.py [hunk 1]
```diff
@@ -8,18 +8,14 @@
 
 
 class FunctionsClient:
-    def __init__(
-        self, url: str, numbers_servers: int = 0, name: Optional[str] = None
-    ):
+    def __init__(self, url: str, numbers_servers: int = 0, name: Optional[str] = None):
```
Comment (andre-senna) [inline hyperon_das/client.py:11]: rename to `server_count`

File: hyperon_das/client.py [hunk 2]
```diff
@@ -32,110 +28,50 @@ def _send_request(self, payload) -> str | dict | int:
         except requests.exceptions.RequestException as e:
             raise e
 
-    def get_node(
-        self,
-        node_type: str,
-        node_name: str,
-        output_format: QueryOutputFormat = QueryOutputFormat.HANDLE,
-    ) -> Union[str, Dict]:
+    def get_node_handle(self, node_type: str, node_name: str) -> str:
```
Comment (andre-senna) [inline hyperon_das/client.py:31]: This is wrong. We don't need to delegate `get_node_handle()` to the remote DAS. This function shouldn't be here. Node and links handles can be calculated using a static method in the basic AtomDB. Se my design for reference.

File: hyperon_das/client.py [hunk 3]
```diff
@@ -32,110 +28,50 @@ def _send_request(self, payload) -> str | dict | int:
...
         }
         return self._send_request(payload)
 
-    def get_nodes(
-        self,
-        node_type: str,
-        node_name: str = None,
-        output_format: QueryOutputFormat = QueryOutputFormat.HANDLE,
-    ) -> Union[List[str], List[Dict]]:
+    def get_link_handle(self, link_type: str, link_targets: List[str]) -> str:
```
Comment (andre-senna) [inline hyperon_das/client.py:38]: This is wrong. We don't need to delegate `get_link_handle()` to the remote DAS. This function shouldn't be here. Node and links handles can be calculated using a static method in the basic AtomDB. Se my design for reference.

File: hyperon_das/das.py [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+
+class DistributedAtomSpace:
+    def __init__(self, **kwargs: Optional[Dict[str, Any]]) -> None:
+        atomdb_parameter = kwargs.get('atomdb', 'ram')
+        query_engine_parameter = kwargs.get('query_engine', 'local')
+
+        if atomdb_parameter == "ram":
+            self.backend = InMemoryDB()
+        elif atomdb_parameter == "redis_mongo":
+            mongo_hostname = kwargs.get('mongo_hostname')
```
Comment (andre-senna) [inline hyperon_das/das.py:35]: These parameters are not supposed to be unpacked here. We should just pass kwargs to MongoDB constructor and let it unpack any parameters it wants.

File: hyperon_das/das.py [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+                redis_hostname=redis_hostname,
+                redis_port=redis_port,
+                redis_username=redis_username,
+                redis_password=redis_password,
+                redis_cluster=redis_cluster,
+                redis_ssl=redis_ssl,
+            )
+            if query_engine_parameter != "local":
+                raise InvalidDASParameters(
+                    message='Can`t instantiate a RedisMongo with query_engine=`local`'
```
Comment (andre-senna) [inline hyperon_das/das.py:74]: The error message doesn't match the `if` check. It seems like the message is the one that's wrong.

File: hyperon_das/das.py [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+                )
+        else:
+            raise ValueError
+            # implement this exception in AtomDB
+            # raise InvalidAtomDB(
+            #    message='This type is an invalid AtomDB', details='Send `ram` or `redis_mongo`'
```
Comment (andre-senna) [inline hyperon_das/das.py:80]: Elglish correction: you shouldn't use the verb`send` to refer to parameters passing. Use `pass` or `use` or `choose` instead.  The verb `send` gives the misleading impression of an asynchronous call. My suggestion: "Invalid AtomDB type. Choose either 'ram' or 'redis_mongo'"

File: hyperon_das/das.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+            >>> print(result)
+            [{'handle': 'dbcf1c7b610a5adea335bf08f6509978', 'type': 'Expression', 'template': ['Expression', 'Symbol', ['Expression', 'Symbol', 'Symbol']], 'targets': [{'handle': '963d66edfb77236054125e3eb866c8b5', 'type': 'Symbol', 'name': 'Test'}, {'handle': '233d9a6da7d49d4164d863569e9ab7b6', 'type': 'Expression', 'template': ['Expression', 'Symbol', 'Symbol'], 'targets': [{'handle': '963d66edfb77236054125e3eb866c8b5', 'type': 'Symbol', 'name': 'Test'}, {'handle': '9f27a331633c8bc3c49435ffabb9110e', 'type': 'Symbol', 'name': '2'}]}]}]
+        """
+        return self.query_engine.query(query, parameters)
+
+    def commit_changes(self):
+        self.query_engine.commit()
+
+    def get_node_handle(self, node_type: str, node_name: str) -> str:
+        return self.backend.node_handle(node_type, node_name)
```
Comment (andre-senna) [inline hyperon_das/das.py:315]: This should be a static method call

File: hyperon_das/das.py [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+
+    def get_link_handle(self, link_type: str, link_targets: List[str]) -> str:
+        return self.backend.link_handle(link_type, link_targets)
```
Comment (andre-senna) [inline hyperon_das/das.py:318]: This should be a static method call

File: hyperon_das/das.py [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+        return self.backend.add_node(node_params)
+
+    def add_link(self, link_params: Dict[str, Any]) -> Dict[str, Any]:
+        return self.backend.add_link(link_params)
+
+    def clear(self) -> None:
+        """Clear all data"""
+        ret = self.backend.clear_database()
+        logger().debug('The database has been cleaned.')
+        return ret
```
Comment (andre-senna) [inline hyperon_das/das.py:330]: This method returns None

**Review by andre-senna — CHANGES_REQUESTED** [review]

File: hyperon_das/das.py [hunk 10]
```diff
@@ -32,52 +28,14 @@ def __init__(self, **kwargs: Optional[Dict[str, Any]]) -> None:
...
-                redis_password=redis_password,
-                redis_cluster=redis_cluster,
-                redis_ssl=redis_ssl,
-            )
+            self.backend = RedisMongoDB(**kwargs)
             if query_engine_parameter != "local":
-                raise InvalidDASParameters(
-                    message='Can`t instantiate a RedisMongo with query_engine=`local`'
-                )
+                raise InvalidDASParameters(message="query_engine parameter must be 'local'")
```
Comment (andre-senna) [inline hyperon_das/das.py:33]: Change messa to "'redis_mongo' backend requires local query engine ('query_engine=local')"

**Review by andre-senna — APPROVED** [review]
Approved but please make the suggested fix before merging

File: hyperon_das/das.py (hunk 10 above)
Comment (marcocapozzoli) [reply hyperon_das/das.py:33]: ok

**Review by andre-senna — APPROVED** [review]

**Pull request merged**